blog/_posts/2020-10-29-load-dynamic.md

```diff
@@ -0,0 +1,126 @@
+the core question of how Bazel got traction at Google, and how it's doing
+outside Google.
+
+I was not on Bazel team. But I remember working with them when they were first
+rolling it out inside Google.
+
+Bazel was a drop-in replacement for gconfig, a Python-based system for
+generating Makefiles.
+
+When Bazel rolled out, here was the plan:
+
+1) Everybody make sure your gconfig BUILD files obey a SUPER strict subset of Python.
+
+2) Bazel will tell you if you did it right, and you can use gconfig or Bazel during the transition.
+
+3) If you did it right, Bazel will run your build WAY faster.
+
+And when I say SUPER strict, as I remember it, that meant only function calls
+were allowed: no local variables, no loops, no `if` branches.
+
+They would have been fine with `load()`, but not with `load_dynamic()`.
+
+But when they tried to roll it out, they hit reality hard. Because there was simply no way
+to express most builds with this strict subset of Python and small feature set.
+
+They introduced a new directive you could add to your config file:
+`PYTHON-PREPROCESSING-REQUIRED`. That told Bazel to let your code use some
+"normal" python. This was a stopgap until Bazel had all the features they needed
+to support complex builds. Then they went back and phased out all the old
+`PYTHON-PREPROCESSING-REQUIRED` stuff.
+
+## The trade-off of load() vs load_dynamic()
+
+Tiltfiles (Tilt's configuration files) are written in
+[Starlark](https://github.com/bazelbuild/starlark), the same underlying
+configuration language as Bazel.
+
+`load()` is a core feature of Starlark. `load_dynamic()` is not.
+
+There's certainly a camp of people that believe supporting imperative Python in
+Bazel was a huge mistake. It made rollout easier, but the pain of phasing it out
+was not worth the benefits of making rollout easier.
+
+I'm not convinced that's true. Early Bazel had a lot of "imperative" features
+that made it easier to adopt, and easier to integrate with other
+tools. Once Google adopted Bazel entirely, the Bazel-team didn't need
+Python pre-processing anymore, and removed it.
+
+<blockquote class="twitter-tweet"><p lang="en" dir="ltr">A periodic &quot;WHYYYYYYYYY BAZELLLLLLL?!&quot;.</p>&mdash; Stephen Augustus (@stephenaugustus) <a href="https://twitter.com/stephenaugustus/status/1318721602050269184?ref_src=twsrc%5Etfw">October 21, 2020</a></blockquote> <script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script>
+
+I don't blame them for removing it! They made Bazel harder to
+optimize, which is critical for a large codebase like Google's.
+
+`load()` vs `load_dynamic()` in Tilt has the same trade-off. 
+
+Long-term, `load()` is easier for us to analyze and optimize. But has some
+limitations that make it harder to experiment with and integrate with other
+tools. In the meantime, we have `load_dynamic()`, so that you can conditionally
+load things, determine what to load at runtime, and more. We're hoping it will
```

Comment:
FWIW:

I read the Bazel story described in this post as:
"even if you know your destination is declarative, it might be useful to have imperative as a stopgap to gain adoption"

When line 117 here says Tilt has the same trade-off, and line 121 says "In the meantime", I read an implication that Tilt is destined for a declarative syntax (which then makes me wonder if you've changed your mind on "The build language of the future will be imperative.") As a user, I'd wonder if this is saying that `load_dynamic` is a stopgap to allow adoption and if it'll go away in the future and make migration work for me.  Or put another way, my best guess as to the answer to "in the meantime until what?" would be "until load_dynamic is removed". Is that a thing you're looking to communicate?

Reply:
ya, this is a good point. I don't think we want to imply that load_dynamic is going away, but rather that you might decide on your own to prefer load over load_dynamic. I rewrote this section a bit.